[dnardnar] iP

dnardnar opened the pull request:
# DNar Bot

> "Your mind is for having ideas, not holding them." – David Allen ([source](https://gettingthingsdone.com))

DNar frees your mind of having to remember things you need to do. It's:

- **Text-based**: Easy to interact with.
- **Easy to learn**: Simple commands.
- **Fast to use**: Quick responses.

All you need to do is:

1. Download it from [here](https://github.com/yourusername/DNar).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😊

And it is **FREE**!

## Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(DNar.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

dnardnar committed:
level-0

dnardnar committed:
level-1

dnardnar committed:
level-1

dnardnar committed:
level-2

dnardnar committed:
level-3

dnardnar committed:
level-4

dnardnar committed:
A-TextUiTesting

dnardnar committed:
level-5

dnardnar committed:
level-6

dnardnar committed:
no message

dnardnar committed:
Merge branch 'branch-Level-7'

dnardnar committed:
no message

dnardnar committed:
Merge branch 'branch-Level-8'

dnardnar committed:
MoreOOP

dnardnar committed:
no message

dnardnar committed:
no message

dnardnar committed:
Merge remote-tracking branch 'origin/add-gradle-support'

dnardnar committed:
no message

dnardnar committed:
no message

dnardnar committed:
no message

dnardnar committed:
A-JavaDoc

dnardnar committed:
Merge branch 'branch-A-JavaDoc'

dnardnar committed:
no message

dnardnar committed:
Merge branch 'branch-A-CodingStandard'

dnardnar committed:
no message

dnardnar committed:
Merge branch 'master' of https://github.com/dnardnar/ip

# Conflicts:
#	src/main/java/dnar/UI.java

farhan-navas commented on src/main/java/dnar/DNar.java:
```diff
@@ -0,0 +1,36 @@
+package dnar;
+
+public class DNar {
```
a few files missing JavaDoc comments, maybe just haven't add yet?

farhan-navas commented on src/main/java/dnar/DNar.java:
```diff
@@ -0,0 +1,36 @@
+
+    public static void main(String[] args) {
+        new DNar("data/DNar.txt").run();
+    }
```
Very good use of encapsulation and oop principles to separate out classes and methods, maybe you could split into different packages too? eg: task, tasklist, todo .. could go into a package called Task and so on..

farhan-navas commented on src/main/java/dnar/UI.java:
```diff
@@ -0,0 +1,153 @@
+ * Handles interactions with the user, including displaying messages
+ * and reading user input.
+ */
+public class UI {
```
Class name should be Ui instead of UI 

[[Link to relevant line in guide]](https://se-education.org/guides/conventions/java/intermediate.html#naming) under the section: Abbreviations and acronyms should not be uppercase when used as a part of a name.

farhan-navas reviewed (commented):
Overall LGTM, just a few coding standard nits to fix!

dnardnar committed:
Update README.md

dnardnar committed:
Update README.md

dnardnar committed:
no message

dnardnar committed:
Merge remote-tracking branch 'origin/master'

dnardnar committed:
Merge branch 'branch-Level-10'

jensenhuangyankai commented on src/main/java/dnar/Task.java:
```diff
@@ -0,0 +1,35 @@
+        this.isDone = isDone;
+    }
+
+    public void markDone() {
```
Good job following naming conventions

jensenhuangyankai commented on src/main/java/dnar/Task.java:
```diff
@@ -0,0 +1,35 @@
+        this.isDone = false;
+    }
+
+    public Task(String description, boolean isDone) {
```
However, consider adding docstrings

jensenhuangyankai commented on src/main/java/dnar/TaskList.java:
```diff
@@ -0,0 +1,102 @@
+     */
+    public void validateIndex(int index) throws DNarException {
+        if (index < 0 || index >= tasks.size()) { // Changed to 0-based index
+            throw new DNarException("This does not exist!! Try 1 to " + tasks.size() + " instead:D");
```
This is not very descriptive to the end user

jensenhuangyankai commented on src/main/java/dnar/Parser.java:
```diff
@@ -0,0 +1,73 @@
+public class Parser {
+    public static void parse(String command, TaskList taskList, UI ui, Storage storage) {
+        try {
+            if (command.equals("list")) {
```
Consider using a switch case for this instead.

dnardnar committed:
Add assertions for code quality

We have introduced Java's built-in `assert` statements to document key assumptions in the code. This helps ensure that certain conditions are met at runtime, improving code reliability and maintainability.

The assertions cover critical points such as:
- Non-null checks for method parameters and objects.
- Index bounds checks for lists and arrays.
- State checks for objects before performing operations.

These assertions are disabled by default and can be enabled with the `-ea` flag during runtime. They serve as a debugging tool to catch unexpected states or programming errors early in development.

dnardnar committed:
Improve code quality across the project
This commit refactors the codebase to enhance readability, maintainability, and adherence to best practices. The changes include:
Readability Improvements:
Avoided long methods by breaking them into smaller, focused functions.
Reduced deep nesting using guard clauses and restructuring logic.
Replaced magic numbers/strings with named constants for clarity.
Simplified complicated expressions by introducing intermediate variables.
Improved logical structure by grouping related statements and adding spacing.
Error-Prone Practices Addressed:
Added default branches in switch statements to handle unexpected cases.
Minimized variable scope and avoided reusing variables for multiple purposes.
Removed dead code and unused parameters to reduce clutter.
Improved Comments:
Enhanced Javadoc comments to explain the purpose (WHAT) and rationale (WHY) of methods and classes.
Avoided redundant comments that repeat obvious information from the code.
These changes aim to make the code easier to read, understand, and modify while reducing potential sources of bugs.

dnardnar committed:
Merge pull request #1 from dnardnar/branch-A-Assertions

Add assertions for code quality

dnardnar committed:
Merge branch 'master' into branch-A-CodeQuality

dnardnar committed:
Merge pull request #2 from dnardnar/branch-A-CodeQuality

Improve code quality across the project

dnardnar committed:
no message

dnardnar committed:
test: Update JUnit tests to ensure compatibility

Refactored and updated JUnit tests to work with the latest changes in the codebase. This improves test coverage and ensures that all functionalities are correctly validated.

dnardnar committed:
docs: Improve Javadoc for main method in DNar application

Enhanced the Javadoc comment for the main method to provide a clearer
and more descriptive explanation of its purpose. The updated comment
follows proper formatting conventions, includes a concise summary, and
clarifies the usage of command line arguments.

This change improves code readability and ensures compliance with
standard documentation practices.

dnardnar committed:
GUI: Connect UI to DNar bot logic

The UI was only echoing user input, not interacting with the bot's core functionality.

Now, the UI calls `Parser.parse()` directly on user input, displaying bot responses in the dialog box.  The DNar class is initialized one time and the methods are called via Parser.

This allows for a proper GUI-driven chatbot experience.

dnardnar committed:
Docs: Add UI screenshot to the docs folder

Adds a representative screenshot of the DNar Chatbot UI to the docs folder. The file is named `Ui.png`, and it shows the full GUI window, with the product name visible in the title bar.

dnardnar committed:
Readme: Revise to serve as a user guide

Revise the readme.md to serve as a comprehensive user guide for the DNar Chatbot. The update includes:
- Detailed instructions on installation and setup.
- Step-by-step usage instructions for common commands.
- Examples of expected input and output.
- Troubleshooting tips for common issues.

This revision aims to enhance user experience and provide clear guidance for new users.

dnardnar committed:
Readme: Update download link and improve clarity

Update the download link in the readme to point to the correct DNar repository. Improve the user guide by providing specific instructions and removing unnecessary brackets from the section headings.

dnardnar committed:
docs: update user guide with new DNar Bot features and commands

- Rewrote the user guide to match the DNar Bot branding and style.
- Added detailed descriptions for all supported commands, including `todo`, `deadline`, `event`, `list`, `mark`, `unmark`, `edit`, `find`, and `delete`.
- Included examples for each command to improve usability.

dnardnar committed:
fix: resolve inconsistencies between UG and application functionality

dnardnar committed:
fix: resolve inconsistencies between UG and application functionality

dnardnar committed:
fix: resolve inconsistencies between UG and application functionality
